kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md

```diff
@@ -0,0 +1,412 @@
+```
+
+Xuddi shu maxsus belgilarni faylga qo'llasak, quyidagicha bo'ladi:
+
+```python
+fayl = open("regifoda.txt", 'r')
+for qator in fayl.readlines():
+    qator = qator.rstrip()
+    lst = re.findall('\S+@\S+', qator)
+    if lst != []:
+        print(lst)
+```
+
+Konsolga
+
+```commandline
+['stephen.marquard@uct.ac.za']
+['<postmaster@collab.sakaiproject.org>']
+['<200801051412.m05ECIaH010327@nakamura.uits.iupui.edu>']
+['<source@collab.sakaiproject.org>;']
+['<source@collab.sakaiproject.org>;']
+['<source@collab.sakaiproject.org>;']
+['apache@localhost)']
+```
+
+Yuqoridagi natija elektron adreslarni chiqarishdir. Lekin, etibor bersak
+bazilarida xar xil biz xoxlamaydigan belgilar mavjud xususan `<;>`. 
+
+> Qanday qilib, ushbu belgilarni olish tashlaymiz? 
+
+Bunda bizga tor'burchak qavs yordamga keladi - `[]` to'rtburchak qavs. Bunda
+biz nimani xoxlashimizni oydinlashtirishimiz mumkin. Masalan, yuqoridagini aytishimiz mumkin
+faqat kichik, yoki katta yoki sonlardan tashkil topsin deb
+
+```python
+[a-zA-Z0-9]
+```
+Ya'ni
+- a-z: a dan z gacha bo'lgan harflar bo'lishi mumkin. Qisqacha hamma kichik harflar
+- A-Z: A dan Z gacha bo'lgan harflar bo'lishi mumkin. Qisqacha hamma Katta harflar
+- 0-9: 0 dan 9 gacha bo'lgan sonlar bo'lishi mumkin
+
+ Shundan foydalin quyidagi programmani yozishimiz mumkin
+
+```python
+tozaroq elektron addresslarni chiqarish
+fayl = open("regifoda.txt", 'r')
+for qator in fayl.readlines():
+    qator = qator.rstrip()
+    lst = re.findall('[a-zA-Z0-9]\S*@\S*[a-zA-Z]', qator)
+    if lst != []:
+        print(lst)
+```
+
+E'tibor bering biz bu yerda `+` o'rniga * ko'paytirish belgisini qo'ydik, chunki `[]` belgisi 1 ta xarf
+o'rniga o'tadi. `*` belgi bittadan ortiq belgilar uchun qo'llaniladi.
+
+Konsolga:
+
+```commandline
+['stephen.marquard@uct.ac.za']
+['postmaster@collab.sakaiproject.org']
+['200801051412.m05ECIaH010327@nakamura.uits.iupui.edu']
+['source@collab.sakaiproject.org']
+['source@collab.sakaiproject.org']
+['source@collab.sakaiproject.org']
+['apache@localhost']
+['source@collab.sakaiproject.org']
+['stephen.marquard@uct.ac.za']
+['source@collab.sakaiproject.org']
+['stephen.marquard@uct.ac.za']
+['stephen.marquard@uct.ac.za']
+['louis@media.berkeley.edu']
+['postmaster@collab.sakaiproject.org']
+['200801042308.m04N8v6O008125@nakamura.uits.iupui.edu']
+['source@collab.sakaiproject.org']
+['source@collab.sakaiproject.org']
+['source@collab.sakaiproject.org']
+['apache@localhost']
+['source@collab.sakaiproject.org']
+```
+
+
+## `()` Qavs
+
+Shunday holatlar bo'ladiki. Biz nafaqat biron nimani qidiramiz
+balki birion malumotni ajratib olishimizga ham to'g'ri keladi. Shunday holatlarda, qavs (`()`) dan 
+foydalanamiz. Keling birinchi quyidagi programmani ishlataylik. Bu yerda maxsus beligilar - 
+`^X\S*: [0-9.]+` shuni anglatadiki: qator boshi `X` bilan boshlansin, va undan keyin bo'shliq bo'lmagan
+bir necha belgilarkelsin toki `:`, va undan keyin ixtoyoriy haqiqiy sonlar kelsin [0-9.]+,
```

Comment:
belgilarkelsin - belgilar kelsin deb ajratilib yozilishi kerak.

Reply:
To'g'irlandi. Rahmat.